Add support for gtk4-session-lock

happenslol opened the pull request:
So, I've generated the files for gtk4-session-lock. I was surprised to see that it actually generates a separate gir file even though both libraries depend on the same lib file in the end (and I suspect the lib can only be imported once because it patches wayland client methods).

Since gir files can't be combined, I've had to create a whole new set of ffi and rust modules/crates here. I think it would be fine to keep these in this repository since they depend on the same library, but if you'd rather I release these myself, I'm fine with that as well.

pentamassiv commented:
Great, that was quick! Thank you!

I had a look at the PR and am happy with your suggestions. I have some minor suggestions and some questions. I didn't want to push to your branch without asking, so I created another branch and added some commits on top of yours. You can find the branch here: https://github.com/pentamassiv/gtk4-layer-shell-gir/tree/pr_suggestions. Of course these commits should all be squashed before merging. I just kept them so you can easily follow my changes. Most should be self-explanatory.

My changes boil down to these points:
- I wanted to generate the code myself to make sure it works as expected
- No manual implementations are needed
- The doc comments can get automatically added as well (using rust-stripper)
- I prefer keeping the example as similar to upstream as possible while writing idiomatic Rust code

Did you manually implement `gtk_session_lock_instance_lock` because you wanted to get a `Result` type as return?

With my changes, the `gtk_session_lock_is_supported` function also gets generated. I added it to the example at the same spot as it is in the upstream example. However this is not working, because [assert_initialized_main_thread!();](https://github.com/pentamassiv/gtk4-layer-shell-gir/commit/ee78f37a966158a125aea53371be831779a675a6#diff-d043a99adfb64a11d1f64a083cf5b22dbfbc3c87e27256363812e52cee8f66d2R18) will panic. Maybe there shouldn't be a call to assert_initialized_main_thread there? If I understand correctly, [gtk gets initialized within the function anyways](https://github.com/wmww/gtk4-layer-shell/blob/cf05a2ba3a00070d22dcd146d6f894507bcdafc9/src/gtk4-session-lock.c#L14). I am not sure where this should be fixed. Can the gir file be changed so that gir won't generate the call to the macro (this will benefit bindings in other languages too) or do we have to manually implement the function (quick and simple)?

When the PR is ready, I would like to reduce it to the parts that were manually written. This will make the PR smaller and the CI will take care of generating the code. Then you also don't have to install rust-stripper to add the doc comments

pentamassiv commented:
I don't mind keeping the two new crates in this repo and I wouldn't mind publishing/maintaining them. I automated keeping the crates up-to-date so I only occasionally have to fix stuff if something breaks. It's likely it also breaks gtk4-layer-shell so I might as well fix it for both crates.
You don't have to feel obligated to add them to this repo though. Feel free to publish them yourself, if you prefer that.
Another option would be to add you as a maintainer and we share the workload

happenslol commented:
Thanks for the feedback! I'll have a look at your branch.

> Did you manually implement gtk_session_lock_instance_lock because you wanted to get a Result type as return?
Yeah, that's exactly it. I thought it would be a good idea to have the interface follow rust conventions as much as possible,, and usually a "success" return value is a `Result<()>`. Even though I like this version better, I can also understand wanting to stick to the original return type.

> With my changes, the gtk_session_lock_is_supported function also gets generated.
Woops! Missed that.

> Maybe there shouldn't be a call to assert_initialized_main_thread there? If I understand correctly, [gtk gets initialized within the function anyways](https://github.com/wmww/gtk4-layer-shell/blob/cf05a2ba3a00070d22dcd146d6f894507bcdafc9/src/gtk4-session-lock.c#L14). I am not sure where this should be fixed. Can the gir file be changed so that gir won't generate the call to the macro (this will benefit bindings in other languages too) or do we have to manually implement the function (quick and simple)?
From what I can see, changing the gir to prevent that is not even possible. I've looked at the [relevant parts of the gir source](https://github.com/gtk-rs/gir/blob/ed4c286a44529251d0e98d8fddfb78838c243127/src/analysis/safety_assertion_mode.rs#L39), and the check is only skipped for methods and specific classes and interfaces (which, by their existence, imply that gtk has already been initialized, I assume).

I suppose manually implementing it might be the only way.

> When the PR is ready, I would like to reduce it to the parts that were manually written. This will make the PR smaller and the CI will take care of generating the code. Then you also don't have to install rust-stripper to add the doc comments
Alright, makes sense.

I'll make your suggested changes, then once we're done with everything, we can remove the generated code :+1:

pentamassiv commented:
Okay, thank you for looking into it. I guess then we can copy the implementation of `gtk_session_lock_is_supported` and remove the call to `assert_initialized_main_thread`. If it's just gir generating them, other bindings won't be affected anyways.

I agree that returning a `Result` would be more idiomatic, but I would prefer sticking to upstream as closely as we can. Generating as much of the code as possible keeps the maintenance effort low and makes it easier for others to audit the code.

pentamassiv commented:
The new crates also need a README.md in their respective folders so users know how they can generate the bindings themselves and we can publish them on crates.io. We can copy most of the existing READMEs from gtk4-layer-shell and gtk4-layer-shell-sys. The main README of the repo needs to be adapted as well.

We can do these changes in a separate PR as well though

happenslol commented:
Aight, pretty sure this is everything we talked about:

- Merged all changes from your branch
- Manually implemented the `is_supported` function to remove the assertion
- Removed all files that are generated by running gir
- Squashed everything into a single commit with a detailed explanation. Could've done multiple commits, but without the generated code it's completely fine I think.

pentamassiv commented:
Could you please also exclude the changes to the files `gtk4-layer-shell-sys/src/lib.rs` and `gtk4-layer-shell-sys/tests/manual.h`?

Can you please copy the doc comments from the generated version of `is_supported` as well? As far as I know rust-stripper can only add doc comments to generated code. The function would be missing it's documentation if we don't manually add it.

After these changes, this PR is good to go :+1: 

After this, we need to add the READMEs, update the publish workflow to be able to publish the new crates and then publish them.

happenslol commented:
> Could you please also exclude the changes to the files gtk4-layer-shell-sys/src/lib.rs and gtk4-layer-shell-sys/tests/manual.h?

I noticed some accidental changes I made to those files and already removed them, you were probably already writing your message when I did that :sweat_smile: 

> Can you please copy the doc comments from the generated version of is_supported as well? As far as I know rust-stripper can only add doc comments to generated code. The function would be missing it's documentation if we don't manually add it.

Weirdly, gir doesn't generate any doc comments for me. I copied the doc comments from your version.

pentamassiv commented:
> I noticed some accidental changes I made to those files and already removed them, you were probably already writing your message when I did that 😅

Did you push all your changes? I still see those files in the commit. I see the doc comments for `is_supported` though.

> Weirdly, gir doesn't generate any doc comments for me. I copied the doc comments from your version.

gir doesn't add them automatically. You would have to run rust-stripper to add them. It is described [here](https://github.com/pentamassiv/gtk4-layer-shell-gir/blob/0b50579bed067982497a08a17e85038d7b2f9819/gtk4-layer-shell/README.md?plain=1#L41), but since you copied it, there's no need for you to run it yourself.

happenslol commented:
> Did you push all your changes? I still see those files in the commit. I see the doc comments for is_supported though.

Yeah, I did, they shouldn't be there in `28a162ff9f7dc22de178ae56b5fad8db2e034830`. I actually just did `git checkout main -- gtk4-layer-shell`/`git checkout main -- gtk4-layer-shell-sys`, so there aren't any changes in there at all.

> gir doesn't add them automatically. You would have to run rust-stripper to add them. It is described [here](https://github.com/pentamassiv/gtk4-layer-shell-gir/blob/0b50579bed067982497a08a17e85038d7b2f9819/gtk4-layer-shell/README.md?plain=1#L41), but since you copied it, there's no need for you to run it yourself.

Ah I see, thanks for the link!

happenslol commented:
> Could you please also exclude the changes to the files gtk4-layer-shell-sys/src/lib.rs and gtk4-layer-shell-sys/tests/manual.h?

Ohh you meant the files in `session-lock`, not `layer-shell`. I removed them as well now. I thought we'd keep the manual file, the lib file was actually just an oversight.

pentamassiv commented:
Woops, sorry about that. Old habits! Yes, I meant those files.

Now the CI is complaining -.-

pentamassiv commented:
I just merged the PR and ran gir myself. This fixed the CI issues. I thought the workflow is capable of generating the code from scratch, but I guess I was wrong.

Thank you for adding the session-lock crates and responding so quickly to my comments. I'll have little time before finishing my exams on Wednesday, so I might respond later or not at all until then. Sorry about that in advance

happenslol commented:
> Thank you for adding the session-lock crates and responding so quickly to my comments. I'll have little time before finishing my exams on Wednesday, so I might respond later or not at all until then. Sorry about that in advance

No problem, thanks from my side as well!

I'll probably find the time to adjust the README and create a new one for the new crates throughout the week.

pentamassiv commented:
Alright, I've published the crates :-)
